tests/test_error_quality.py

```
@@ -0,0 +1,222 @@
+    Before: `raise ValueError(f"Operation {i}: ...")` — the index was only
+    in the message string. Worse, `_apply_single_edit` raised bare
+    `ValueError("replace requires 'find' and 'replace_with'")` with NO
+    index at all. On a 10-op batch, the agent could not tell which op
+    failed without re-running operations one by one.
+
+    After: `BatchOperationError(DocxEditError)` with `.operation_index`
+    and `.reason`. Every validation path — pre-dispatch and inside
+    `_apply_single_edit` — is wrapped so the index is always present.
+    """
+
+    def _build_batch_doc(self, doc_path):
```

Ruff (0.15.10)

[warning] Missing return type annotation for private function `_build_batch_doc`

(ANN202)